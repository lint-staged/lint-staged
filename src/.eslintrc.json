{
  "plugins": ["node"],
  "extends": [
    "plugin:node/recommended"
  ],
  "rules": {
    "node/exports-style": [
      "error",
      "module.exports"
    ]
  }
}
