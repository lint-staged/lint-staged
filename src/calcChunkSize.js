'use strict'

/**
 * Calculates and returns the chunk size for given file paths and `chunkSize`
 * option.
 *
 * It returns the minimum of the following:
 *
 *  - Total number of files
 *  - Max allowed chunk size so that command length does not exceed the system
 *    limitation on windows
 *  - User specified chunk size or the default
 *
 * Worked example:
 * **Assumption** - Our max file path length is 100, Hence max allowed chunk
 * size is 80
 *
 *  - Case 1: Only 10 files are there, chunk size should be 10 only
 *  - Case 2: There are 100 files and user has overridden the option with
 *    chunk size 40. So chunk size should be 40
 *  - Case 3: There are 100 files and user has overridden the option with
 *    chunk size 100. So chunk size should be 80
 *
 * @param {Array<string>} paths The array of file paths
 * @param {number} idealChunkSize User specified / default chunk size
 * @returns {number} The chunk size
 */
module.exports = function calcChunkSize(paths, idealChunkSize) {
    /* What is the longest file path? */
    const maxPathLen = paths.reduce(
        (maxLen, filePath) => Math.max(maxLen, filePath.length),
        20 // safe initial value
    )

    /* In the worst case scenario, */
    /* how many files can we process in a single command? */
    /* For windows systems, command length is limited to 8192 */
    const maxAllowedChunkSize = Math.floor(8000 / maxPathLen)

    /* Configured chunk size / default - idealChunkSize */
    return Math.min(paths.length, maxAllowedChunkSize, idealChunkSize)
}
