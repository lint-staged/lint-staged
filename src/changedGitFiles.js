const { exec } = require('child_process')

module.exports = function changedGitFiles(opMode, callback) {
    const gitDiffCmd = 'git diff-tree -r --name-only --no-commit-id'
    let revA = 'HEAD@{1}'
    let revB = 'HEAD'

    if (opMode === 'checkout') {
        const params = process.env.GIT_PARAMS.split(' ')

        if (params[2] === '0') {
            // Exit early if this was only a file checkout, not a branch change ($3 == 1)
            callback(null, [])
            return
        }

        revA = params[0]
        revB = params[1]
    }

    exec(`${ gitDiffCmd } ${ revA } ${ revB }`, (err, sources) => {
        if (err) {
            callback(err)
        } else {
            callback(null, sources.split('\n').map(filename => ({ filename })))
        }
    })
}
