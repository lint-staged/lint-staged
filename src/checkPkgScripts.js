'use strict'

const chalk = require('chalk')
const dedent = require('dedent')
const has = require('lodash/has')

const warn = msg => {
  console.warn(chalk.yellowBright.bold(msg))
}

/**
 * Checks if the given command or binary name is present in the package.json scripts. This would be
 * called if and when resolving a binary fails in `findBin`.
 *
 * @param {Object} pkg package.json
 * @param {string} cmd
 * @param {string} binName
 * @param {Array<string>} args
 * @throws {Error} If a script is found in the pkg for the given `cmd` or `binName`.
 */
module.exports = function checkPkgScripts(pkg, cmd, binName, args) {
  if (pkg && pkg.scripts) {
    const { scripts } = pkg
    let scriptName
    let script
    if (has(scripts, cmd)) {
      scriptName = cmd
      script = scripts[cmd]
    } else if (has(scripts, binName)) {
      scriptName = binName
      script = scripts[binName]
    } else {
      return
    }

    const argsStr = args && args.length ? args.join(' ') : ''
    warn(dedent`
      \`lint-staged\` no longer supports running scripts defined in package.json.

      The same behavior can be achieved by changing the command to any of the following:
        - \`npm run ${scriptName} -- ${argsStr}\`
        - \`${script} ${argsStr}\`
    `)
    throw new Error(`Could not resolve binary for \`${cmd}\``)
  }
}
