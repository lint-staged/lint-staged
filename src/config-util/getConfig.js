'use strict'

const intersection = require('lodash/intersection')
const defaultsDeep = require('lodash/defaultsDeep')
const isObject = require('lodash/isObject')
const defaultConfig = require('./defaultConfig').defaultConfig

/**
 * Check if the config is "simple" i.e. doesn't contains any of full config keys.
 *
 * This check is not concerned with the validity of the config. For example,
 * en empty object or array will satisfy the check.
 *
 * @param config
 * @returns {boolean}
 */
function isSimple(config) {
  /*
    Scenarios which satisfy `isSimple` check:
    - []
    - {}
    - [
        {
          "includes": ["src/*.js"],
          "excludes": ["src/*.ignore.js"],
          "commands": ["eslint --fix", "git add"]
        }
      ]
    - { "*.js": ["eslint --fix", "git add"] }

    Scenario which do not satisfy the check:
    - {
        "linters": { "*.js": ["eslint --fix", "git add"] }
      }
    - {
        "linters": []
      }
  */
  return (
    Array.isArray(config) ||
    (isObject(config) &&
      !config.hasOwnProperty('linters') &&
      intersection(Object.keys(defaultConfig), Object.keys(config)).length === 0)
  )
}

function asArray(value) {
  if (Array.isArray(value)) return value
  return [value]
}

/**
 * Expand shorthand object format to the array format. If `linters` is an array, it is
 * returned as is. If it is a plain object, it is converted from
 * `{ "*.js": "eslint --fix" }` to
 * `[ { "includes": ["*.js"], "excludes": [], "commands": "eslint --fix" } ]`
 *
 * @param {Array|Object|null|undefined} sourceConfig
 * @returns {Array<Object>}
 */
function expandShorthands(isSimpleConfig, sourceConfig) {
  if (sourceConfig == null) return []

  const linters = isSimpleConfig ? sourceConfig : sourceConfig.linters
  if (linters == null) return []

  if (Array.isArray(linters)) {
    return linters.map(linter => ({
      includes: linter.includes || [],
      excludes: linter.excludes || [],
      commands: asArray(linter.commands)
    }))
  }

  return Object.keys(linters).map(pattern => ({
    includes: [pattern],
    excludes: [],
    commands: asArray(linters[pattern])
  }))
}

/**
 * For a given configuration object that we retrive from .lintstagedrc or package.json
 * construct a full configuration with all options set.
 *
 * This is a bit tricky since we support 2 different syntxes: simple and full
 * For simple config, only the `linters` configuration is provided.
 *
 * @param {Object} sourceConfig
 * @returns {{
 *  concurrent: boolean, chunkSize: number, gitDir: string, globOptions: {matchBase: boolean, dot: boolean}, linters: {}, subTaskConcurrency: number, renderer: string, verbose: boolean
 * }}
 */
module.exports = function getConfig(sourceConfig) {
  const isSimpleConfig = isSimple(sourceConfig)
  const config = defaultsDeep(
    // Do not mutate sourceConfig!!!
    { linters: expandShorthands(isSimpleConfig, sourceConfig) },
    isSimpleConfig ? null : sourceConfig,
    defaultConfig
  )

  // Check if renderer is set in sourceConfig and if not, set accordingly to verbose
  if (isObject(sourceConfig) && !sourceConfig.hasOwnProperty('renderer')) {
    config.renderer = config.verbose ? 'verbose' : 'update'
  }

  return config
}
