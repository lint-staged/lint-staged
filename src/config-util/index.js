'use strict'

exports.getConfig = require('./getConfig')
exports.validateConfig = require('./validateConfig')
