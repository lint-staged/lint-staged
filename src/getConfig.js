/* eslint no-prototype-builtins: 0 */

'use strict'

const chalk = require('chalk')
const format = require('stringify-object')
const intersection = require('lodash/intersection')
const defaultsDeep = require('lodash/defaultsDeep')
const isObject = require('lodash/isObject')
const validate = require('jest-validate').validate
const logValidationWarning = require('jest-validate').logValidationWarning
const unknownOptionWarning = require('jest-validate/build/warnings').unknownOptionWarning
const isGlob = require('is-glob')

/**
 * Default config object
 *
 * @type {{concurrent: boolean, chunkSize: number, gitDir: string, globOptions: {matchBase: boolean, dot: boolean}, linters: {}, subTaskConcurrency: number, renderer: string, verbose: boolean}}
 */
const defaultConfig = {
  concurrent: true,
  chunkSize: Number.MAX_SAFE_INTEGER,
  gitDir: '.',
  globOptions: {
    matchBase: true,
    dot: true
  },
  linters: {},
  subTaskConcurrency: 1,
  renderer: 'update',
  verbose: false
}

/**
 * Check if the config is "simple" i.e. doesn't contains any of full config keys
 *
 * @param config
 * @returns {boolean}
 */
function isSimple(config) {
  return (
    isObject(config) &&
    !config.hasOwnProperty('linters') &&
    intersection(Object.keys(defaultConfig), Object.keys(config)).length === 0
  )
}

/**
 * Custom jest-validate reporter for unknown options
 * @param config
 * @param example
 * @param option
 * @param options
 * @returns {void}
 */
function unknownValidationReporter(config, example, option, options) {
  /**
   * If the unkonwn property is a glob this is probably
   * a typical mistake of mixing simple and advanced configs
   */
  if (isGlob(option)) {
    const message = `  Unknown option ${chalk.bold(`"${option}"`)} with value ${chalk.bold(
      format(config[option], { inlineCharacterLimit: Number.POSITIVE_INFINITY })
    )} was found in the config root.

  You are probably trying to mix simple and advanced config formats. Adding

  ${chalk.bold(`"linters": {
    "${option}": ${JSON.stringify(config[option])}
  }`)}

  will fix it and remove this message.`

    const comment = options.comment
    const name = (options.title && options.title.warning) || 'WARNING'
    return logValidationWarning(name, message, comment)
  }
  // If it is not glob pattern, use default jest-validate reporter
  return unknownOptionWarning(config, example, option, options)
}

/**
 * For a given configuration object that we retrive from .lintstagedrc or package.json
 * construct a full configuration with all options set.
 *
 * This is a bit tricky since we support 2 different syntxes: simple and full
 * For simple config, only the `linters` configuration is provided.
 *
 * @param {Object} sourceConfig
 * @returns {{
 *  concurrent: boolean, chunkSize: number, gitDir: string, globOptions: {matchBase: boolean, dot: boolean}, linters: {}, subTaskConcurrency: number, renderer: string, verbose: boolean
 * }}
 */
function getConfig(sourceConfig) {
  const config = defaultsDeep(
    {}, // Do not mutate sourceConfig!!!
    isSimple(sourceConfig) ? { linters: sourceConfig } : sourceConfig,
    defaultConfig
  )

  // Check if renderer is set in sourceConfig and if not, set accordingly to verbose
  if (isObject(sourceConfig) && !sourceConfig.hasOwnProperty('renderer')) {
    config.renderer = config.verbose ? 'verbose' : 'update'
  }

  return config
}

/**
 * Runs config validation. Throws error if the config is not valid.
 * @param config {Object}
 * @returns config {Object}
 */
function validateConfig(config) {
  const exampleConfig = Object.assign({}, defaultConfig, {
    linters: {
      '*.js': ['eslint --fix', 'git add'],
      '*.css': 'stylelint'
    }
  })

  const validation = validate(config, {
    exampleConfig,
    unknown: unknownValidationReporter,
    comment:
      'Please refer to https://github.com/okonet/lint-staged#configuration for more information...'
  })

  if (!validation.isValid) {
    throw new Error('lint-staged config is invalid... Aborting.')
  }

  return config
}

module.exports = {
  getConfig,
  validateConfig
}
