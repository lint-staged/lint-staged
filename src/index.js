/* eslint no-console: 0 */
/* eslint import/no-dynamic-require: 0 */

'use strict'

const appRoot = require('app-root-path')
const cosmiconfig = require('cosmiconfig')
const stringifyObject = require('stringify-object')
const getConfig = require('./getConfig').getConfig
const validateConfig = require('./getConfig').validateConfig
const printErrors = require('./printErrors')
const runAll = require('./runAll')

// Find the right package.json at the root of the project
// TODO: Test if it should be aware of `gitDir`
const packageJson = require(appRoot.resolve('package.json'))

// Force colors for packages that depend on https://www.npmjs.com/package/supports-color
// but do this only in TTY mode
if (process.stdout.isTTY) {
  process.env.FORCE_COLOR = true
}

const errConfigNotFound = new Error('Config could not be found')

/**
 * Root lint-staged function that is called from .bin
 */
module.exports = function lintStaged() {
  return cosmiconfig('lint-staged', {
    rc: '.lintstagedrc',
    rcExtensions: true
  })
    .then(result => {
      if (result == null) throw errConfigNotFound

      // result.config is the parsed configuration object
      // result.filepath is the path to the config file that was found
      const config = validateConfig(getConfig(result.config))

      if (config.verbose) {
        console.log(`
Running lint-staged with the following config:
${stringifyObject(config)}
`)
      }

      runAll(packageJson, config)
        .then(() => {
          // No errors, exiting with 0
          process.exitCode = 0
        })
        .catch(error => {
          // Errors detected, printing and exiting with non-zero
          printErrors(error)
          process.exitCode = 1
        })
    })
    .catch(err => {
      if (err === errConfigNotFound) {
        console.error(`${err.message}.`)
      } else {
        // It was probably a parsing error
        console.error(`Could not parse lint-staged config.

${err}`)
      }
      // Print helpful message for all errors
      console.error(`
Please make sure you have created it correctly.
See https://github.com/okonet/lint-staged#configuration.
`)
      process.exitCode = 1
    })
}
