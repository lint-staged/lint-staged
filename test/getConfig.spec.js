/* eslint no-console: 0 */

import { makeConsoleMock } from 'consolemock'
import cloneDeep from 'lodash/cloneDeep'
import { getConfig, validateConfig } from '../src/getConfig'

describe('getConfig', () => {
  it('should return config with defaults for undefined', () => {
    expect(getConfig()).toMatchSnapshot()
  })

  it('should return config with defaults', () => {
    expect(getConfig({})).toMatchSnapshot()
  })

  it('should set verbose', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        verbose: false
      })
    )

    expect(
      getConfig({
        verbose: false
      })
    ).toEqual(
      expect.objectContaining({
        verbose: false
      })
    )

    expect(
      getConfig({
        verbose: true
      })
    ).toEqual(
      expect.objectContaining({
        verbose: true
      })
    )
  })

  it('should set concurrent', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        concurrent: true
      })
    )

    expect(
      getConfig({
        concurrent: false
      })
    ).toEqual(
      expect.objectContaining({
        concurrent: false
      })
    )

    expect(
      getConfig({
        concurrent: true
      })
    ).toEqual(
      expect.objectContaining({
        concurrent: true
      })
    )
  })

  it('should set renderer based on verbose key', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        renderer: 'update'
      })
    )

    expect(
      getConfig({
        '*.js': ['eslint', 'git add']
      })
    ).toEqual(
      expect.objectContaining({
        renderer: 'update'
      })
    )

    expect(
      getConfig({
        verbose: false
      })
    ).toEqual(
      expect.objectContaining({
        renderer: 'update'
      })
    )

    expect(
      getConfig({
        verbose: true
      })
    ).toEqual(
      expect.objectContaining({
        renderer: 'verbose'
      })
    )
  })

  it('should set gitDir', () => {
    expect(getConfig({})).toEqual(
      expect.objectContaining({
        gitDir: '.'
      })
    )

    expect(
      getConfig({
        gitDir: '../path'
      })
    ).toEqual(
      expect.objectContaining({
        gitDir: '../path'
      })
    )
  })

  it('should set linters', () => {
    expect(getConfig()).toEqual(
      expect.objectContaining({
        linters: {}
      })
    )

    expect(getConfig({})).toEqual(
      expect.objectContaining({
        linters: {}
      })
    )

    expect(
      getConfig({
        '*.js': 'eslint'
      })
    ).toEqual(
      expect.objectContaining({
        linters: {
          '*.js': 'eslint'
        }
      })
    )

    expect(
      getConfig({
        linters: {
          '*.js': ['eslint --fix', 'git add'],
          '.*rc': 'jsonlint'
        }
      })
    ).toMatchSnapshot()
  })

  it('should deeply merge configs', () => {
    expect(
      getConfig({
        globOptions: {
          nocase: true
        }
      })
    ).toEqual(
      expect.objectContaining({
        globOptions: {
          nocase: true,
          matchBase: true,
          dot: true
        }
      })
    )
  })

  it('should not add plain linters object to the full config', () => {
    expect(
      getConfig({
        '*.js': 'eslint'
      })
    ).not.toEqual(
      expect.objectContaining({
        '*.js': 'eslint'
      })
    )
  })

  it('should not change config if the whole config was passed', () => {
    const src = {
      concurrent: false,
      chunkSize: 2,
      gitDir: '/to',
      globOptions: {
        matchBase: false,
        dot: true
      },
      linters: {
        '*.js': 'eslint'
      },
      subTaskConcurrency: 10,
      renderer: 'custom',
      verbose: true
    }
    expect(getConfig(cloneDeep(src))).toEqual(src)
  })
})

describe('validateConfig', () => {
  beforeEach(() => {
    global.console = makeConsoleMock()
  })

  it('should throw and should print validation errors for invalid config', () => {
    const invalidAdvancedConfig = {
      foo: false,
      chunkSize: 'string',
      gitDir: 111
    }
    expect(() => validateConfig(getConfig(invalidAdvancedConfig))).toThrowErrorMatchingSnapshot()
  })

  it('should not throw and should print validation warnings for mixed config', () => {
    const invalidMixedConfig = {
      gitDir: './path/to/packagejson/',
      '*.js': ['eslint --fix', 'git add']
    }
    expect(() => validateConfig(getConfig(invalidMixedConfig))).not.toThrow()
    expect(console.printHistory()).toMatchSnapshot()
  })

  it('should not throw and should print nothing for simple valid config', () => {
    const validSimpleConfig = {
      '*.js': ['eslint --fix', 'git add']
    }
    expect(() => validateConfig(getConfig(validSimpleConfig))).not.toThrow()
    expect(console.printHistory()).toMatchSnapshot()
  })

  it('should not throw and should print nothing for advanced valid config', () => {
    const validAdvancedConfig = {
      gitDir: '.',
      linters: {
        '*.js': ['eslint --fix', 'git add']
      }
    }
    expect(() => validateConfig(getConfig(validAdvancedConfig))).not.toThrow()
    expect(console.printHistory()).toMatchSnapshot()
  })
})
