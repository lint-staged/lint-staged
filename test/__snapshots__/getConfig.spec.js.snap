// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConfig should return config with defaults 1`] = `
Object {
  "chunkSize": 9007199254740991,
  "concurrent": true,
  "gitDir": ".",
  "globOptions": Object {
    "dot": true,
    "matchBase": true,
  },
  "linters": Object {},
  "renderer": "update",
  "subTaskConcurrency": 1,
  "verbose": false,
}
`;

exports[`getConfig should return config with defaults for undefined 1`] = `
Object {
  "chunkSize": 9007199254740991,
  "concurrent": true,
  "gitDir": ".",
  "globOptions": Object {
    "dot": true,
    "matchBase": true,
  },
  "linters": Object {},
  "renderer": "update",
  "subTaskConcurrency": 1,
  "verbose": false,
}
`;

exports[`getConfig should set linters 1`] = `
Object {
  "chunkSize": 9007199254740991,
  "concurrent": true,
  "gitDir": ".",
  "globOptions": Object {
    "dot": true,
    "matchBase": true,
  },
  "linters": Object {
    "*.js": Array [
      "eslint --fix",
      "git add",
    ],
    ".*rc": "jsonlint",
  },
  "renderer": "update",
  "subTaskConcurrency": 1,
  "verbose": false,
}
`;

exports[`validateConfig should not throw and should print nothing for advanced valid config 1`] = `""`;

exports[`validateConfig should not throw and should print nothing for simple valid config 1`] = `""`;

exports[`validateConfig should not throw and should print validation warnings for mixed config 1`] = `
"
WARN ● Validation Warning:

  Unknown option \\"*.js\\" with value ['eslint --fix', 'git add'] was found in the config root.

  You are probably trying to mix simple and advanced config formats. Adding

  \\"linters\\": {
    \\"*.js\\": [\\"eslint --fix\\",\\"git add\\"]
  }

  will fix it and remove this message.

Please refer to https://github.com/okonet/lint-staged#configuration for more information..."
`;

exports[`validateConfig should throw and should print validation errors for invalid config 1`] = `
"● Validation Error:

  Option \\"chunkSize\\" must be of type:
    number
  but instead received:
    string

  Example:
  {
    \\"chunkSize\\": 9007199254740991
  }

Please refer to https://github.com/okonet/lint-staged#configuration for more information..."
`;
